File: java/src/tree/BinarySearchTree.java
package tree;

public class BinarySearchTree
{
    private Node _root;

    public Node getRoot()
    {
        return this._root;
    }

    public void add(int value)
    {
        if (this._root == null) {
            this._root = new Node(value);
            return;
        }

        Node node = this._root;
        while (true) {
            if (value > node.getValue()) {
                if (node.getRight() != null) {
                    node = node.getRight();
                    continue;
                }
                node.setRightNode(value);
            }
            else {
                if (node.getLeft() != null) {
                    node = node.getLeft();
                    continue;
                }
                node.setLeftNode(value);
            }
            break;
        }
    }

    public int find(int value)
    {
        Node node = getRoot();

        if (node == null) {
            return -1;
        }

        int iteration = 0;
        while (true) {
            if (value == node.getValue()) {
                return iteration;
            }

            if (value > node.getValue() && node.getRight() != null) {
                node = node.getRight();
                iteration++;
                continue;
            }
            else {
                if (node.getLeft() != null) {
                    node = node.getLeft();
                    iteration++;
                    continue;
                }
            }

            if (node.getRight() == null) {
                break;
            }
        }
        return -1;
    }
}

File: cpp/src/BinarySearchTree.cpp

#include <memory>
#include <iostream>
#include "include/BinarySearchTree.hpp"

std::shared_ptr<Node> BinarySearchTree::root() const
{
    return m_root;
}

void BinarySearchTree::add(int value)
{
    if (!m_root)
    {
        m_root = std::make_shared<Node>(value);
        return;
    }

    std::shared_ptr<Node> node = root();
    while (true)
    {
        if (value > node->value)
        {
            if (node->right != nullptr)
            {
                node = node->right;
                continue;
            }
            node->right = std::make_shared<Node>(value);
            break;
        }
        else
        {
            if (node->left != nullptr)
            {
                node = node->left;
                continue;
            }

            node->left = std::make_shared<Node>(value);
            break;
        }
    }
}

int BinarySearchTree::find(int value) const
{
    std::shared_ptr<Node> node = this->root();
    if (node == nullptr)
    {
        return -1;
    }

    int iteration = 0;
    while (true)
    {
        if (value == node->value)
        {
            return iteration;
        }

        iteration++;

        if (value > node->value && node->right != nullptr)
        {
            node = node->right;
            continue;
        }
        else
        {
            if (node->left != nullptr)
            {
                node = node->left;
                continue;
            }
        }

        break;
    }
    return -1;
}
